packages/doenetml-worker-rust/lib-doenetml-core/src/new_core/prop_calculation.rs

```diff
@@ -0,0 +1,119 @@
+use crate::{
+    components::{
+        prelude::{ComponentState, DataQuery, FlatDastElementContent},
+        ComponentNode,
+    },
+    graph::directed_graph::Taggable,
+    state::{Freshness, PropPointer},
+    ComponentIdx,
+};
+
+use super::graph_based_core::Core;
+
+impl Core {
+    pub fn freshen_renderer_state(&mut self) -> Vec<ComponentIdx> {
+        let mut stale_renderers = Vec::new();
+
+        // recursively get a list of all rendered descendants of the components in stale_renderers
+        let mut stale_renderer_idx = 0;
+        while stale_renderer_idx < stale_renderers.len() {
+            let component_idx = stale_renderers[stale_renderer_idx];
+            stale_renderer_idx += 1;
+
+            stale_renderers.extend(
+                self.component_rendered_children(&self.components[component_idx])
+                    .into_iter()
+                    .filter_map(|child| match child {
+                        FlatDastElementContent::Text(_) => None,
+                        FlatDastElementContent::Element(child_idx) => Some(child_idx),
+                    }),
+            );
+        }
```

Comment:
Most of our components are getting rendered each time, right? You could just do a `Vec<bool>` the length of `self.components` and mark things as stale that way. Then no dedup is required.

Reply:
Typically, only a tiny subset are being rerendered each time (at least as far as core is concerned). It would just be the components who have a direct dependency on the props being changed by an action. E.g., when typing in an input box, the input itself is typically the only component being rerendered, unless something is dependent on its `immediateValue` prop.

For the initial render, it is a different story, though even then many components don't render their children, components in attributes aren't rendered, and we have tags like `<setup>` for computations that aren't rendered.

I think for the typical small updates where speed is of concern (like when dragging a point), we want an algorithm that scales with the number of components updated, not the total number of components. Other than that, I'm not sure what is the most efficient approach.